[RLlib] Fix MultiAgentEpisode.get_infos() to include __common__ key

yashwantbezawada opened the pull request:
Fixes #58668

## Summary

The `MultiAgentEpisode.get_infos()` method was excluding the `__common__` key from the infos dictionary returned to the caller. This key is used in multi-agent environments to store information that is common to all agents.

## Problem

When an environment's `step()` method returns infos with a `__common__` key like this:

```python
infos = {
    "player1": {"player1_data": "foo"},
    "player2": {"player2_data": "bar"},
    "__common__": {"common_data": "baz"},
}
```

The `__common__` key was being lost because:
1. In `add_env_step()`, only agent IDs appearing in observations, actions, rewards, or terminateds/truncateds were processed
2. The `__common__` key only appears in infos, so it was never included in the `agent_ids_with_data` set
3. As a result, `__common__` infos were silently dropped and not stored anywhere

## Solution

This fix follows the same pattern used for the `__all__` key in terminateds/truncateds:

1. **Added storage**: Added `_hanging_infos_end` field to `__slots__` and initialized it in `__init__`
2. **Extract on add**: In `add_env_step()`, extract and store the `__common__` key from the infos dict before processing agent-specific infos
3. **Include on get**: Modified all three retrieval code paths to include the `__common__` key in the returned dictionary:
   - `_get_data_by_agent_steps()` (when `env_steps=False`)
   - `_get_data_by_env_steps()` (default case)
   - `_get_data_by_env_steps_as_list()` (when `return_list=True`)

## Testing

The fix can be verified with the reproduction script from the issue. After this change, calling `episode.get_infos()` will correctly include the `__common__` key in the returned dictionary.

## Changes

- Added `_hanging_infos_end` to `__slots__`
- Initialized `_hanging_infos_end = {}` in `__init__()`
- Extract `__common__` from infos in `add_env_step()` and store it in `_hanging_infos_end`
- Include `__common__` in returned dict from all `get_infos()` code paths

yashwantbezawada committed:
[RLlib] Fix MultiAgentEpisode.get_infos() to include __common__ key

Fixes #58668

The MultiAgentEpisode.get_infos() method was excluding the __common__ key
from the infos dictionary. This key is used in multi-agent environments
to store information that is common to all agents.

Changes:
- Add _hanging_infos_end field to store infos including __common__
- Extract __common__ from infos dict in add_env_step()
- Include __common__ in returned dict from all get_infos() code paths:
  - _get_data_by_agent_steps()
  - _get_data_by_env_steps()
  - _get_data_by_env_steps_as_list()

The fix follows the same pattern used for __all__ key in terminateds/truncateds.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
Add serialization and concatenation support for _hanging_infos_end

- Add _hanging_infos_end to get_state() for proper state serialization
- Add _hanging_infos_end restoration in from_state() with backward compatibility
- Copy __common__ infos in concat_episode() when concatenating episode chunks

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
[RLlib] Fix __common__ info loss in cut() and constructor

This commit addresses two related bugs where the __common__ key in
infos would be lost during episode operations:

1. **cut() method**: When cutting an episode to create a successor chunk,
   the _hanging_infos_end dictionary (containing __common__) was not being
   copied to the successor. This caused __common__ info to be lost when
   episodes were cut. Now we copy _hanging_infos_end["__common__"] to the
   successor, similar to how _hanging_rewards_end is handled.

2. **_init_single_agent_episodes() method**: When creating episodes via the
   constructor with infos containing __common__, this key was being silently
   dropped during per-agent data extraction. The method only extracted
   agent-specific infos (e.g., inf.get(agent_id, {})) but ignored __common__.
   Now we extract and preserve __common__ in _hanging_infos_end.

These fixes ensure __common__ infos are properly preserved across episode
cutting and construction workflows, addressing the concerns raised in
automated code review.

Refs: #58668
Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

pseudo-rnd-thoughts commented:
Thanks for the PR @yashwantbezawada, could you implement a test for this.

pseudo-rnd-thoughts commented on rllib/env/multi_agent_episode.py:
```diff
@@ -250,6 +251,7 @@ def __init__(
         # until the next observation is received.
         self._hanging_actions_end = {}
         self._hanging_extra_model_outputs_end = defaultdict(dict)
+        self._hanging_infos_end = {}
```
@yashwantbezawada Could you make these fixes and include them in the tests

yashwantbezawada replied:
concerns were addressed in commit 51f6581

yashwantbezawada committed:
[RLlib] Add test for __common__ key preservation in get_infos()

This test verifies that the __common__ key is properly preserved when
calling MultiAgentEpisode.get_infos() with different parameters.

The test covers all three code paths:
- _get_data_by_env_steps() (env_steps=True, default)
- _get_data_by_agent_steps() (env_steps=False)
- _get_data_by_env_steps_as_list() (return_list=True)

Addresses maintainer feedback on PR #58673.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
[RLlib] Add incremental test for __common__ key preservation

This test addresses the maintainer's request to have a test that builds
an episode incrementally using add_env_step(), simulating how episodes
are actually created in practice. The test verifies that the __common__
key in infos is preserved and updated correctly as the episode grows.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
[RLlib] Fix add_env_reset() to extract __common__ infos

This addresses the Cursor bot feedback about add_env_reset() not extracting
the __common__ key from infos. The method now extracts and stores __common__
infos before processing agent-specific infos, matching the pattern used in
add_env_step(). This ensures __common__ infos are preserved throughout the
episode lifecycle, including during reset operations.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
[RLlib] Use pop() in _init_single_agent_episodes for consistency

Address Cursor bot feedback: use inf.pop("__common__") instead of
inf["__common__"] to be consistent with add_env_reset() and add_env_step().
This ensures the __common__ key is properly removed from the original dict,
preventing memory bloat and maintaining behavioral consistency across all
methods that handle __common__ infos.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>
